### PR TITLE
Nix flake

### PR DESCRIPTION
**Pull request by Gurjaka**
This is a fully functional Qtile flake for Nixos. It is used to pull latest commits/updates on nixos. Jwijenbergh was working on a flake before, but I tried to clean it up a bit, and make my first contribution to Qtile this way. 

Special thanks to "Yohann Boniface/Sigmanificient", I couldn't have done this without his help! :)))

image

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        });
+    in {
+      default = self.packages.${pkgs.system}.qtile;
+      qtile = (override-qtile pypkgs.qtile);
+    });
+
+    devShells = forAllSystems (pkgs: let
+      common-shell = {
+        env = {
+          QTILE_DLOPEN_LIBGOBJECT = "${pkgs.glib}/lib/libgobject-2.0.so.0";
```
Comment (jwijenbergh) [inline flake.nix:38]: These env vars don't do anything now, for `nix develop` to work you need: https://github.com/jwijenbergh/qtile/commit/e312855cddb20a123a965f40f4b3094995f5d937 (the parts where it patches our dlopen calls)

Hence why I haven't PRed this myself yet as I don't know a good solution to this problem

File: nix/overlays.nix [hunk 2]
```diff
@@ -0,0 +1,40 @@
+self: final: prev: {
+  pythonPackagesOverlays =
+    (prev.pythonPackagesOverlays or [])
+    ++ [
+      (_: pprev: {
+        pywlroots = (pprev.pywlroots.overrideAttrs(_: rec {
```
Comment (jwijenbergh) [inline nix/overlays.nix:6]: Pywlroots has been upstreamed, the overlay can be removed

File: nix/overlays.nix [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            rev = "bf314bae5d7a9552225c44ab1e7bf7e4fafa869e";
+            hash = "sha256-tvem8gmih9EfseZ3MgU1UW05xfttzKLcJEaymM5uNPI=";
+          };
+        })).override {
+          wlroots = prev.wlroots_0_17;
+        };
+        qtile = (pprev.qtile.overrideAttrs (old: let
+          flakever = self.shortRev or "dev";
+        in {
+          version = "0.0.0+${flakever}.flake";
```
Comment (jwijenbergh) [inline nix/overlays.nix:20]: We need a way to deduce the version number, maybe you can parse it from the changelog?

File: nix/overlays.nix (hunk 2 above)
Comment (Gurjaka) [reply nix/overlays.nix:6]: I have overlay just in case, when qtile moves on to 0.18

File: flake.nix (hunk 1 above)
Comment (jwijenbergh) [reply flake.nix:38]: See also that commit I linked how to replace GitHub actions with nix :)

File: nix/overlays.nix (hunk 2 above)
Comment (jwijenbergh) [reply nix/overlays.nix:6]: We can just add it back when that happens

File: flake.nix (hunk 1 above)
Comment (jwijenbergh) [reply flake.nix:38]: Some test fixes from that commit are also good to pull in here

File: nix/overlays.nix (hunk 2 above)
Comment (Gurjaka) [reply nix/overlays.nix:6]: Why delete tho?

Comment (jwijenbergh) [reply nix/overlays.nix:6]: I don't like to have unnecessary code around

File: nix/overlays.nix (hunk 3 above)
Comment (Gurjaka) [reply nix/overlays.nix:20]: I don't understand what you want me to do, can you explain better, what do you want me to parse from changelog?

File: nix/overlays.nix (hunk 2 above)
Comment (Gurjaka) [reply nix/overlays.nix:6]: Well, we can remove it ig

**Comment by jwijenbergh** [comment]
Thanks for picking up this! I added some comments and I have a general question here:

Any reason you removed the development shells for each python version? That was good for testing. Also can you add a git workflow that tests the nix build?

File: nix/overlays.nix (hunk 3 above)
Comment (jwijenbergh) [reply nix/overlays.nix:20]: Now it says 0.0.0+commit.flake,

It should say 0.28.1.flake for a release buikd

Comment (Gurjaka) [reply nix/overlays.nix:20]: Ohh i get it, so the version + flake

**Comment by Gurjaka** [comment]
> Thanks for picking up this! I added some comments and I have a general question here:
> 
> Any reason you removed the development shells for each python version? That was good for testing. Also can you add a git workflow that tests the nix build?

File: nix/overlays.nix (hunk 3 above)
Comment (jwijenbergh) [reply nix/overlays.nix:20]: Yeah

Comment (jwijenbergh) [reply nix/overlays.nix:20]: For dev/git commits that are not tagged yet 0.0.0 or so is fine

**Comment by Gurjaka** [comment]
Accidentally closed and reopened, so removed development shells for each python version, since I didn't think it was necessary :D. And about git workflow, I will try to do that, but rn my brain isn't braining, I can barely even write this comment, maybe we can discuss this further tomorrow in discord?

File: flake.nix (hunk 1 above)
Comment (Sigmanificient) [reply flake.nix:38]: we can fix this using
```nix
        env.LD_LIBRARY_PATH = pkgs.lib.makeLibraryPath [
          pkgs.glib
          pkgs.pango
          pkgs.pango
          pkgs.xcb-util-cursor
          pkgs.pixman
          pkgs.libdrm.dev
        ];
```
but patching will be required for cffi build script

**Comment by Sigmanificient** [comment]
image

**Comment by jwijenbergh** [comment]
Been a while since I made a mistake with the git cli hah!

**Review by jwijenbergh — APPROVED** [review]

**Comment by jwijenbergh** [comment]
@elParaguayo any thoughts on this?

**Comment by jwijenbergh** [comment]
fine to be merged by @elParaguayo. Do you see any downsides of adding this to the repo? @tych0 

This is mainly useful for Nix(OS) users that either:
- want to contribute to qtile
- use a git version of qtile in their config

It could be a separate repo but that is normally not done with Nix and it complicates keeping them in sync

We have to update these files when:
- we add a new dependency or need to pin a certain dependency
- periodically we have to run `nix flake update` to ensure that the `flake.lock` is kept up to date. I could make a CI action for this

**Comment by tych0** [comment]
I don't have a problem with it, although,

> periodically we have to run nix flake update to ensure that the flake.lock is kept up to date. I could make a CI action for this

Yeah, I wonder if this should be part of the release action maybe? Or whenever someone remembers to run it (maybe via a make target)?

Also, is there any way to keep the multiple files out of /? We already have a lot of random packaging trash there, it would be nice to not add more.

**Comment by jwijenbergh** [comment]
> I don't have a problem with it, although,
> 
> > periodically we have to run nix flake update to ensure that the flake.lock is kept up to date. I could make a CI action for this
> 
> Yeah, I wonder if this should be part of the release action maybe? Or whenever someone remembers to run it (maybe via a make target)?

Yeah, do you want to me to add something to the release action in this pr or is it ok that I improve on it later? A make target is easier but it's essentially just running `nix flake update`

> Also, is there any way to keep the multiple files out of /? We already have a lot of random packaging trash there, it would be nice to not add more.

There is a way, however, it's a bit unconventional with nix. instead or running `nix develop` you would run `nix develop subdir`. I think most users are used to the default location of the flake being the root

**Comment by tych0** [comment]
> Yeah, do you want to me to add something to the release action in this pr or is it ok that I improve on it later? A make target is easier but it's essentially just running nix flake update

I would just put it in a make target in this PR, and that's good enough for me.

> There is a way, however, it's a bit unconventional with nix. instead or running nix develop you would run nix develop subdir. I think most users are used to the default location of the flake being the root

Ok, no worries. Most of the trash is because of python's terrible packaging anyway, what's a little bit of Nix :)

**Comment by jwijenbergh** [comment]
> > Yeah, do you want to me to add something to the release action in this pr or is it ok that I improve on it later? A make target is easier but it's essentially just running nix flake update
> 
> I would just put it in a make target in this PR, and that's good enough for me.
> 
> > There is a way, however, it's a bit unconventional with nix. instead or running nix develop you would run nix develop subdir. I think most users are used to the default location of the flake being the root
> 
> Ok, no worries. Most of the trash is because of python's terrible packaging anyway, what's a little bit of Nix :)

I added a makefile target and updated the flake again with the command

**Comment by jwijenbergh** [comment]
@tych0 can I merge this?

**Comment by tych0** [comment]
Sure! I went ahead :)

**Pull request merged**